src/dist/mod.rs

```diff
@@ -146,12 +145,80 @@ pub struct PartialToolchainDesc {
 /// 1.55-x86_64-pc-windows-msvc
 #[derive(Debug, Clone, Eq, PartialEq, PartialOrd, Ord)]
 pub struct ToolchainDesc {
-    // Either "nightly", "stable", "beta", or an explicit version number
-    pub channel: String,
+    pub channel: ToolchainChannel,
     pub date: Option<String>,
     pub target: TargetTriple,
 }
 
+#[derive(Debug, Clone, Eq, PartialEq, PartialOrd, Ord)]
+pub enum ToolchainChannel {
+    Stable,
+    Beta,
+    Nightly,
+    Ver(PartialVersion),
+}
+
+impl fmt::Display for ToolchainChannel {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Stable => write!(f, "stable"),
+            Self::Beta => write!(f, "beta"),
+            Self::Nightly => write!(f, "nightly"),
+            Self::Ver(ver) => write!(f, "{ver}"),
+        }
+    }
+}
+
+impl FromStr for ToolchainChannel {
+    type Err = anyhow::Error;
+    fn from_str(chan: &str) -> Result<Self> {
+        match chan {
+            "stable" => Ok(Self::Stable),
+            "beta" => Ok(Self::Beta),
+            "nightly" => Ok(Self::Nightly),
+            ver => ver.parse().map(Self::Ver),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, PartialOrd, Ord)]
+pub struct PartialVersion {
+    pub major: u64,
```

Comment:
Would suggest using `u8`, `u16`, `u8` for these to use minimal types.

Reply:
@djc This is to align with https://docs.rs/semver/latest/semver/struct.Comparator.html#'s field declarations to simplify the translation from `Comparator` to `PartialVersion`. Should we truncate it explicitly?

Reply:
Hmm, I suppose it might not be worth it.